src/book.rs

```diff
@@ -189,6 +189,17 @@ impl Book {
 
         // match until either the order is fully matched, or all valid orders to match with are done.
         while !done {
+            /* prevent self-matching */
+            if orders_queue.front().is_some()
```

Comment:
Does `unwrap()` pop the order from the queue? 

If it is, then we don't put it back on the queue after the continue. But if it isn't popping, then how does the call on line 204 to `orders_queue.pop_front()` not pick up this order?

The tests pass so I'm obviously missing something here but just wanted to check my understanding

Reply:
@raymogg [`front`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.front) simply provides an *immutable* reference to the element at the front of the deque, or `None` if the deque is empty. All [`unwrap`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap) does is retrieve the element within the `Option` returned by `front` and panics if the option is `None`.

To avoid a panic, the first term of the conditional bounds checks the length of the deque, which, due to short-circuit evaluation of boolean expressions, will skip the `unwrap` call.